fw/http.c

```diff
@@ -2677,23 +2704,48 @@ tfw_http_expand_hbh(TfwHttpResp *resp, unsigned short status)
 static int
 tfw_http_set_hdr_connection(TfwHttpMsg *hm, unsigned long conn_flg)
 {
+	int r;
 	BUILD_BUG_ON(BIT_WORD(__TFW_HTTP_MSG_M_CONN) != 0);
 	if (((hm->flags[0] & __TFW_HTTP_MSG_M_CONN) == conn_flg)
 	    && (!TFW_STR_EMPTY(&hm->h_tbl->tbl[TFW_HTTP_HDR_CONNECTION]))
-	    && !test_bit(TFW_HTTP_B_CONN_EXTRA, hm->flags))
+	    && !test_bit(TFW_HTTP_B_CONN_EXTRA, hm->flags)
+	    && !test_bit(TFW_HTTP_B_CONN_UPGRADE, hm->flags))
+	{
 		return 0;
+	}
 
-	switch (conn_flg) {
-	case BIT(TFW_HTTP_B_CONN_CLOSE):
+	if (unlikely(conn_flg == BIT(TFW_HTTP_B_CONN_CLOSE)))
 		return TFW_HTTP_MSG_HDR_XFRM(hm, "Connection", "close",
 					     TFW_HTTP_HDR_CONNECTION, 0);
-	case BIT(TFW_HTTP_B_CONN_KA):
-		return TFW_HTTP_MSG_HDR_XFRM(hm, "Connection", "keep-alive",
-					     TFW_HTTP_HDR_CONNECTION, 0);
-	default:
-		return TFW_HTTP_MSG_HDR_DEL(hm, "Connection",
-					    TFW_HTTP_HDR_CONNECTION);
+
+	if (conn_flg == BIT(TFW_HTTP_B_CONN_KA)) {
+		if (test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET, hm->flags)
+		    && test_bit(TFW_HTTP_B_CONN_UPGRADE, hm->flags))
+		{
+			r = TFW_HTTP_MSG_HDR_XFRM(hm, "Connection",
+						  "keep-alive, upgrade",
+						  TFW_HTTP_HDR_CONNECTION, 0);
+		}
+		else {
+			r = TFW_HTTP_MSG_HDR_XFRM(hm, "Connection",
+						  "keep-alive",
+						  TFW_HTTP_HDR_CONNECTION, 0);
+		}
+	} else {
+		if (test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET, hm->flags)
+		    && test_bit(TFW_HTTP_B_CONN_UPGRADE, hm->flags))
+		{
+			r = TFW_HTTP_MSG_HDR_XFRM(hm, "Connection",
+						  "upgrade",
+						  TFW_HTTP_HDR_CONNECTION, 0);
```

Comment:
If we have `Connection: close, upgrade`, then it seems we should send just `Connection: close`. Also could you please check whether a connection is considered keep-alive by default (there is neither `close` nor `keep-alive`)?

Reply:
By default passed 0 in the function for responses, `BIT(TFW_HTTP_B_CONN_KA)` passed to function for responses only if paired request has `TFW_HTTP_B_CONN_KA` bit.

But for requests by default passed `BIT(TFW_HTTP_B_CONN_KA)`.

And semantically connection is keep-alive by default for `HTTP/1.1` and later.

Reply:
If we have `Connection: close, upgrade` in response than now result depends on status. If status is `4XX` than connection is closed without upgrade. If `101` than upgrade is finished and connection state than depends on actual closing on no-closing from backend.

May be that is right enough.